File: packages/conformance/src/expect.ts
import type { EvaluationContext, HeaderExpectation, ObservedResponse, StepExpect } from './types';

const decoder = new TextDecoder();

function bytesEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.byteLength !== b.byteLength) return false;
  for (let i = 0; i < a.byteLength; i++) {
    if (a[i] !== b[i]) return false;
  }
  return true;
}

function checkHeader(
  name: string,
  expectation: HeaderExpectation,
  headers: Headers,
): string | undefined {
  const actual = headers.get(name);
  if (typeof expectation === 'string') {
    if (actual === null) return `header ${name}: expected "${expectation}", got absent`;
    if (actual !== expectation) return `header ${name}: expected "${expectation}", got "${actual}"`;
    return undefined;
  }
  if ('present' in expectation) {
    return actual === null ? `header ${name}: expected present, got absent` : undefined;
  }
  if ('absent' in expectation) {
    return actual === null ? undefined : `header ${name}: expected absent, got "${actual}"`;
  }
  const re = new RegExp(expectation.regex);
  if (actual === null) return `header ${name}: expected /${expectation.regex}/, got absent`;
  if (!re.test(actual)) return `header ${name}: expected /${expectation.regex}/, got "${actual}"`;
  return undefined;
}

function checkBodyEquals(
  expected: string | { sameAs: string },
  body: Uint8Array,
  ctx: EvaluationContext,
): string | undefined {
  if (typeof expected === 'string') {
    const actual = decoder.decode(body);
    return actual === expected ? undefined : `body: expected "${expected}", got "${actual}"`;
  }
  const prior = ctx.priorBodies.get(expected.sameAs);
  if (prior === undefined) return `body: sameAs references unknown step ${expected.sameAs}`;
  if (bytesEqual(prior, body)) return undefined;
  return `body: expected same bytes as step ${expected.sameAs} (${prior.byteLength} bytes), got ${body.byteLength} bytes that differ`;
}

function checkBodyJson(expected: Record<string, unknown>, body: Uint8Array): string[] {
  let parsed: unknown;
  try {
    parsed = JSON.parse(decoder.decode(body));
  } catch {
    return ['body: expected JSON object, got unparseable body'];
  }
  if (typeof parsed !== 'object' || parsed === null || Array.isArray(parsed)) {
    return ['body: expected JSON object, got non-object'];
  }
  const record = parsed as Record<string, unknown>;
  const failures: string[] = [];
  for (const [key, value] of Object.entries(expected)) {
    if (record[key] !== value) {
      failures.push(
        `body.${key}: expected ${JSON.stringify(value)}, got ${JSON.stringify(record[key])}`,
      );
    }
  }
  return failures;
}

/** Returns a list of human readable failures; empty means the observed response satisfies the expectation. */
export function evaluateExpect(
  expect: StepExpect,
  observed: ObservedResponse,
  ctx: EvaluationContext,
): string[] {
  const failures: string[] = [];
  if (observed.status !== expect.status) {
    failures.push(`status: expected ${expect.status}, got ${observed.status}`);
  }
  for (const [name, expectation] of Object.entries(expect.headers ?? {})) {
    const failure = checkHeader(name, expectation, observed.headers);
    if (failure) failures.push(failure);
  }
  if (expect.bodyEquals !== undefined) {
    const failure = checkBodyEquals(expect.bodyEquals, observed.body, ctx);
    if (failure) failures.push(failure);
  }
  if (expect.bodyJson !== undefined) {
    failures.push(...checkBodyJson(expect.bodyJson, observed.body));
  }
  if (expect.bodyBytes !== undefined && observed.body.byteLength !== expect.bodyBytes) {
    failures.push(`body: expected ${expect.bodyBytes} bytes, got ${observed.body.byteLength}`);
  }
  if (expect.handlerInvocations !== undefined) {
    if (ctx.handlerInvocations === undefined) {
      failures.push('handlerInvocations: counter unavailable');
    } else if (ctx.handlerInvocations !== expect.handlerInvocations) {
      failures.push(
        `handlerInvocations: expected ${expect.handlerInvocations}, got ${ctx.handlerInvocations}`,
      );
    }
  }
  return failures;
}

File: go/conformance/expect.go
package conformance

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"regexp"
	"sort"
)

type observed struct {
	status int
	header http.Header
	body   []byte
}

type evalContext struct {
	prior       map[string][]byte
	invocations *int
}

func jsonText(v any) string {
	b, err := json.Marshal(v)
	if err != nil {
		return fmt.Sprint(v)
	}
	return string(b)
}

func checkHeader(name string, exp HeaderExpectation, h http.Header) string {
	actual, present := h[http.CanonicalHeaderKey(name)]
	value := ""
	if present {
		value = actual[0]
	}
	switch {
	case exp.Exact != nil:
		if !present {
			return fmt.Sprintf("header %s: expected %q, got absent", name, *exp.Exact)
		}
		if value != *exp.Exact {
			return fmt.Sprintf("header %s: expected %q, got %q", name, *exp.Exact, value)
		}
	case exp.Present:
		if !present {
			return fmt.Sprintf("header %s: expected present, got absent", name)
		}
	case exp.Absent:
		if present {
			return fmt.Sprintf("header %s: expected absent, got %q", name, value)
		}
	case exp.Regex != "":
		re, err := regexp.Compile(exp.Regex)
		if err != nil {
			return fmt.Sprintf("header %s: invalid regex /%s/", name, exp.Regex)
		}
		if !present {
			return fmt.Sprintf("header %s: expected /%s/, got absent", name, exp.Regex)
		}
		if !re.MatchString(value) {
			return fmt.Sprintf("header %s: expected /%s/, got %q", name, exp.Regex, value)
		}
	}
	return ""
}

// evaluate mirrors packages/conformance/src/expect.ts message for message.
func evaluate(exp StepExpect, obs observed, ctx evalContext) []string {
	failures := []string{}
	if obs.status != exp.Status {
		failures = append(failures, fmt.Sprintf("status: expected %d, got %d", exp.Status, obs.status))
	}
	names := make([]string, 0, len(exp.Headers))
	for name := range exp.Headers {
		names = append(names, name)
	}
	sort.Strings(names)
	for _, name := range names {
		if f := checkHeader(name, exp.Headers[name], obs.header); f != "" {
			failures = append(failures, f)
		}
	}
	if exp.BodyEquals != nil {
		switch {
		case exp.BodyEquals.Exact != nil:
			if string(obs.body) != *exp.BodyEquals.Exact {
				failures = append(failures, fmt.Sprintf("body: expected %q, got %q", *exp.BodyEquals.Exact, string(obs.body)))
			}
		default:
			prior, ok := ctx.prior[exp.BodyEquals.SameAs]
			if !ok {
				failures = append(failures, fmt.Sprintf("body: sameAs references unknown step %s", exp.BodyEquals.SameAs))
			} else if !bytes.Equal(prior, obs.body) {
				failures = append(failures, fmt.Sprintf("body: expected same bytes as step %s (%d bytes), got %d bytes that differ", exp.BodyEquals.SameAs, len(prior), len(obs.body)))
			}
		}
	}
	if exp.BodyJSON != nil {
		var parsed any
		if err := json.Unmarshal(obs.body, &parsed); err != nil {
			failures = append(failures, "body: expected JSON object, got unparseable body")
		} else if object, ok := parsed.(map[string]any); !ok {
			failures = append(failures, "body: expected JSON object, got non-object")
		} else {
			keys := make([]string, 0, len(exp.BodyJSON))
			for k := range exp.BodyJSON {
				keys = append(keys, k)
			}
			sort.Strings(keys)
			for _, k := range keys {
				got := "undefined"
				if value, present := object[k]; present {
					got = jsonText(value)
				}
				if got != jsonText(exp.BodyJSON[k]) {
					failures = append(failures, fmt.Sprintf("body.%s: expected %s, got %s", k, jsonText(exp.BodyJSON[k]), got))
				}
			}
		}
	}
	if exp.BodyBytes != nil && len(obs.body) != *exp.BodyBytes {
		failures = append(failures, fmt.Sprintf("body: expected %d bytes, got %d", *exp.BodyBytes, len(obs.body)))
	}
	if exp.HandlerInvocations != nil {
		switch {
		case ctx.invocations == nil:
			failures = append(failures, "handlerInvocations: counter unavailable")
		case *ctx.invocations != *exp.HandlerInvocations:
			failures = append(failures, fmt.Sprintf("handlerInvocations: expected %d, got %d", *exp.HandlerInvocations, *ctx.invocations))
		}
	}
	return failures
}
